File: src/main/java/minegame159/fireball/parser/Scanner.java
package minegame159.fireball.parser;

import java.io.IOException;
import java.io.Reader;

public class Scanner {
    private final Reader reader;
    private final StringBuilder sb = new StringBuilder();

    private char current, next;
    private int line = 1, character, firstTokenCharacter;

    private boolean inString;

    public Scanner(Reader reader) {
        this.reader = reader;

        advance(false);
        advance(false);
    }

    public Token next() {
        sb.setLength(0);

        skipWhitespace();
        if (isAtEnd()) return token(TokenType.Eof);

        firstTokenCharacter = character;
        char c = advance();

        if (c == 'c' && peek() == '{') {
            advance(false);
            return cBlock();
        }
        if (isDigit(c)) return number();
        if (isAlpha(c)) return identifier();

        return switch (c) {
            case '(' -> token(TokenType.LeftParen);
            case ')' -> token(TokenType.RightParen);
            case '{' -> token(TokenType.LeftBrace);
            case '}' -> token(TokenType.RightBrace);
            case ':' -> token(TokenType.Colon);
            case ';' -> token(TokenType.Semicolon);
            case ',' -> token(TokenType.Comma);
            case '.' -> token(TokenType.Dot);

            case '!' -> token(match('=') ? TokenType.BangEqual : TokenType.Bang);
            case '=' -> token(match('=') ? TokenType.EqualEqual : TokenType.Equal);
            case '>' -> token(match('=') ? TokenType.GreaterEqual : TokenType.Greater);
            case '<' -> token(match('=') ? TokenType.LessEqual : TokenType.Less);

            case '+' -> token(match('=') ? TokenType.PlusEqual : TokenType.Plus);
            case '-' -> token(match('=') ? TokenType.MinusEqual : TokenType.Minus);
            case '*' -> token(match('=') ? TokenType.StarEqual : TokenType.Star);
            case '/' -> token(match('=') ? TokenType.SlashEqual : TokenType.Slash);
            case '%' -> token(match('=') ? TokenType.PercentageEqual : TokenType.Percentage);

            case '&' -> token(match('&') ? TokenType.And : TokenType.Ampersand);
            case '|' -> token(match('|') ? TokenType.Or : TokenType.Pipe);

            case '"' -> string();

            default -> error("Unexpected character");
        };
    }

    private Token cBlock() {
        boolean countTrimStart = true;
        int trimStart = 0;
        
        int braceCount = 0;

        while (!isAtEnd()) {
            if (countTrimStart) {
                if (isMeaninglessWhitespace(peek())) trimStart++;
                else countTrimStart = false;
            }

            if (peek() == '\n') line++;
            else if (peek() == '{') braceCount++;
            else if (peek() == '}') {
                if (braceCount > 0) braceCount--;
                else break;
            }

            advance();
        }

        if (isAtEnd()) return error("Unterminated C block.");

        advance(false);
        sb.delete(0, trimStart + 1);

        int trimEnd = 0;
        for (int i = sb.length() - 1; i >= 0; i--) {
            if (isMeaninglessWhitespace(sb.charAt(i))) trimEnd++;
            else break;
        }
        sb.delete(sb.length() - trimEnd, sb.length() + trimEnd);

        return new Token(TokenType.CBlock, sb.toString(), line, firstTokenCharacter);
    }

    private Token number() {
        while (isDigit(peek())) advance();

        if (peek() == '.' && isDigit(peekNext())) {
            advance();

            while (isDigit(peek())) advance();

            return token(TokenType.Float);
        }

        return token(TokenType.Int);
    }

    private Token string() {
        inString = true;
        while (peek() != '"' && !isAtEnd()) {
            if (peek() == '\n') line++;
            advance();
        }

        if (isAtEnd()) return error("Unterminated string.");

        advance();
        inString = false;

        return token(TokenType.String);
    }

    private Token identifier() {
        while (isAlpha(peek()) || isDigit(peek())) advance();
        return token(identifierType());
    }

    private TokenType identifierType() {
        return switch (sb.charAt(0)) {
            case 'n' -> checkKeyword(1, "ull", TokenType.Null);
            case 't' -> checkKeyword(1, "rue", TokenType.True);
            case 'f' -> {
                if (sb.length() > 1) yield switch (sb.charAt(1)) {
                    case 'a' -> checkKeyword(2, "lse", TokenType.False);
                    case 'o' -> checkKeyword(2, "r", TokenType.For);
                    default -> TokenType.Identifier;
                };
                yield TokenType.Identifier;
            }
            case 'i' -> checkKeyword(1, "f", TokenType.If);
            case 'w' -> checkKeyword(1, "hile", TokenType.While);
            case 'v' -> checkKeyword(1, "ar", TokenType.Var);
            case 'e' -> checkKeyword(1, "lse", TokenType.Else);
            case 'r' -> checkKeyword(1, "eturn", TokenType.Return);
            case 's' -> checkKeyword(1, "truct", TokenType.Struct);

            default -> TokenType.Identifier;
        };
    }

    private TokenType checkKeyword(int start, String rest, TokenType type) {
        return sb.substring(start).equals(rest) ? type : TokenType.Identifier;
    }

    // Utils

    private boolean match(char expected) {
        if (isAtEnd()) return false;
        if (peek() != expected) return false;

        advance();
        return true;
    }

    private char advance() {
        return advance(true);
    }

    private char advance(boolean append) {
        if (append) sb.append(current);

        char prev = current;
        current = next;

        try {
            int i = reader.read();
            next = i == -1 ? '\0' : (char) i;
        } catch (IOException e) {
            e.printStackTrace();
        }

        if (prev != '\0' && (inString || prev != '\n')) character++;
        return prev;
    }

    private char peek() {
        return current;
    }

    private char peekNext() {
        return next;
    }

    private boolean isAtEnd() {
        return peek() == '\0';
    }

    private boolean isMeaninglessWhitespace(char c) {
        return c == ' ' || c == '\r' || c == '\t';
    }

    private void skipWhitespace() {
        while (true) {
            if (isMeaninglessWhitespace(peek())) {
                advance(false);
                continue;
            }

            switch (peek()) {
                case '\n' -> {
                    incrementLine();
                    advance(false);
                }
                case '/' -> {
                    if (peekNext() == '/') {
                        while (peek() != '\n' && !isAtEnd()) advance(false);
                    }
                    else return;
                }
                default -> {
                    return;
                }
            }
        }
    }

    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    private boolean isAlpha(char c) {
        return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
    }

    private void incrementLine() {
        line++;
        character = 0;
    }

    private Token token(TokenType type) {
        return new Token(type, sb.toString(), line, firstTokenCharacter);
    }

    private Token error(String msg) {
        return new Token(TokenType.Error, msg, line, firstTokenCharacter);
    }
}

File: core/scanner/scanner.go
package scanner

import (
	"fireball/core"
)

type Scanner struct {
	text string

	startI   int
	currentI int

	line   uint16
	column uint16

	startLine   uint16
	startColumn uint16
}

func NewScanner(text string) *Scanner {
	return &Scanner{
		text:     text,
		startI:   0,
		currentI: 0,
		line:     1,
	}
}

func (s *Scanner) Next() PositionedToken {
	if token := s.skipWhitespace(); !token.IsEmpty() {
		return token
	}

	s.start()

	if s.isAtEnd() {
		return s.make(Eof)
	}

	c := s.advance()

	if isAlpha(c) {
		return s.identifier()
	}
	if isDigit(c) || (c == '-' && isDigit(s.peek())) {
		return s.number(c)
	}

	switch c {
	case '(':
		return s.make(LeftParen)
	case ')':
		return s.make(RightParen)
	case '{':
		return s.make(LeftBrace)
	case '}':
		return s.make(RightBrace)
	case '[':
		return s.make(LeftBracket)
	case ']':
		return s.make(RightBracket)

	case '.':
		if s.peek() == '.' && s.peekNext() == '.' {
			s.advance()
			s.advance()

			return s.make(DotDotDot)
		}

		return s.make(Dot)
	case ',':
		return s.make(Comma)
	case ':':
		return s.make(Colon)
	case ';':
		return s.make(Semicolon)

	case '+':
		if s.match('+') {
			return s.make(PlusPlus)
		}

		return s.matchToken('=', PlusEqual, Plus)
	case '-':
		if s.match('-') {
			return s.make(MinusMinus)
		}

		return s.matchToken('=', MinusEqual, Minus)
	case '*':
		return s.matchToken('=', StarEqual, Star)
	case '/':
		return s.matchToken('=', SlashEqual, Slash)
	case '%':
		return s.matchToken('=', PercentageEqual, Percentage)

	case '=':
		if s.match('=') {
			return s.make(EqualEqual)
		}

		return s.matchToken('>', FuncPtr, Equal)
	case '!':
		return s.matchToken('=', BangEqual, Bang)
	case '<':
		if s.match('<') {
			return s.matchToken('=', LessLessEqual, LessLess)
		}

		return s.matchToken('=', LessEqual, Less)
	case '>':
		if s.match('>') {
			return s.matchToken('=', GreaterGreaterEqual, GreaterGreater)
		}

		return s.matchToken('=', GreaterEqual, Greater)

	case '|':
		if s.match('=') {
			return s.make(PipeEqual)
		}

		return s.matchToken('|', Or, Pipe)
	case '^':
		if s.match('=') {
			return s.make(XorEqual)
		}

		return s.make(Xor)
	case '&':
		if s.match('=') {
			return s.make(AmpersandEqual)
		}

		return s.matchToken('&', And, Ampersand)

	case '#':
		return s.make(Hashtag)

	case '\'':
		return s.character()
	case '"':
		return s.string()
	}

	return s.error("Unexpected character.")
}

func (s *Scanner) identifier() PositionedToken {
	for isAlpha(s.peek()) || isDigit(s.peek()) {
		s.advance()
	}

	return s.make(s.identifierKind())
}

func (s *Scanner) identifierKind() TokenKind {
	switch s.text[s.startI] {
	case 'a':
		return s.checkKeyword(1, "s", As)
	case 'b':
		return s.checkKeyword(1, "reak", Break)
	case 'c':
		return s.checkKeyword(1, "ontinue", Continue)
	case 'e':
		if s.currentI-s.startI > 1 {
			switch s.text[s.startI+1] {
			case 'l':
				return s.checkKeyword(2, "se", Else)
			case 'n':
				return s.checkKeyword(2, "um", Enum)
			}
		}
	case 'f':
		if s.currentI-s.startI > 1 {
			switch s.text[s.startI+1] {
			case 'a':
				return s.checkKeyword(2, "lse", False)
			case 'n':
				return s.checkKeyword(2, "", Fn)
			case 'o':
				return s.checkKeyword(2, "r", For)
			case 'u':
				return s.checkKeyword(2, "nc", Func)
			}
		}
	case 'i':
		if s.currentI-s.startI > 1 {
			switch s.text[s.startI+1] {
			case 'f':
				return s.checkKeyword(2, "", If)
			case 'm':
				return s.checkKeyword(2, "pl", Impl)
			case 'n':
				return s.checkKeyword(2, "terface", Interface)
			case 's':
				return s.checkKeyword(2, "", Is)
			}
		}
	case 'n':
		if s.currentI-s.startI > 1 {
			switch s.text[s.startI+1] {
			case 'a':
				return s.checkKeyword(2, "mespace", Namespace)
			case 'i':
				return s.checkKeyword(2, "l", Nil)
			}
		}
	case 'r':
		return s.checkKeyword(1, "eturn", Return)
	case 's':
		if s.currentI-s.startI > 1 {
			switch s.text[s.startI+1] {
			case 't':
				if s.currentI-s.startI > 2 {
					switch s.text[s.startI+2] {
					case 'a':
						return s.checkKeyword(3, "tic", Static)
					case 'r':
						return s.checkKeyword(3, "uct", Struct)
					}
				}
			}
		}
	case 't':
		return s.checkKeyword(1, "rue", True)
	case 'u':
		return s.checkKeyword(1, "sing", Using)
	case 'v':
		return s.checkKeyword(1, "ar", Var)
	case 'w':
		return s.checkKeyword(1, "hile", While)
	}

	return Identifier
}

func (s *Scanner) checkKeyword(start int, rest string, kind TokenKind) TokenKind {
	if s.currentI-s.startI == start+len(rest) && s.text[s.startI+start:s.startI+start+len(rest)] == rest {
		return kind
	}

	return Identifier
}

func (s *Scanner) number(c uint8) PositionedToken {
	next := s.peek()

	// Hex
	if c == '0' && (next == 'x' || next == 'X') {
		s.advance()
		return s.hex()
	}

	// Binary
	if c == '0' && (next == 'b' || next == 'B') {
		s.advance()
		return s.binary()
	}

	// Integers or floats
	return s.integerOrFloat()
}

func (s *Scanner) integerOrFloat() PositionedToken {
	for isDigit(s.peek()) {
		s.advance()
	}

	if s.peek() == '.' && isDigit(s.peekNext()) {
		s.advance()

		for isDigit(s.peek()) {
			s.advance()
		}
	}

	if s.peek() == 'f' || s.peek() == 'F' {
		s.advance()
	}

	return s.make(Number)
}

func (s *Scanner) hex() PositionedToken {
	for isHex(s.peek()) {
		s.advance()
	}

	return s.make(Hex)
}

func (s *Scanner) binary() PositionedToken {
	for isBinary(s.peek()) {
		s.advance()
	}

	return s.make(Binary)
}

func (s *Scanner) string() PositionedToken {
	for s.peek() != '"' && !s.isAtEnd() {
		if s.peek() == '\n' {
			s.line++
		}

		s.advance()
	}

	if s.isAtEnd() {
		return s.error("Unterminated string")
	}

	s.advance()
	return s.make(String)
}

func (s *Scanner) character() PositionedToken {
	if s.isAtEnd() || s.peek() == '\'' {
		return s.error("Empty character.")
	}

	if s.advance() == '\\' && !s.isAtEnd() {
		c := s.advance()

		if c != '\'' && c != '0' && c != 'n' && c != 'r' && c != 't' {
			return s.error("Unexpected character.")
		}
	}

	if s.peek() != '\'' {
		return s.error("Unterminated character.")
	}

	s.advance()
	return s.make(Character)
}

func (s *Scanner) matchToken(expected uint8, kindTrue TokenKind, kindFalse TokenKind) PositionedToken {
	if s.match(expected) {
		return s.make(kindTrue)
	}

	return s.make(kindFalse)
}

func (s *Scanner) match(expected uint8) bool {
	if s.isAtEnd() {
		return false
	}

	if s.peek() != expected {
		return false
	}

	s.advance()
	return true
}

func (s *Scanner) skipWhitespace() PositionedToken {
	for {
		if s.isAtEnd() {
			return PositionedToken{}
		}

		c := s.peek()

		switch c {
		case ' ', '\r', '\t':
			s.advance()

		case '\n':
			s.advance()
			s.line++
			s.column = 0

		case '/':
			if s.peekNext() == '/' {
				s.start()

				for !s.isAtEnd() && s.peek() != '\n' {
					s.advance()
				}

				return s.make(Comment)
			} else if s.peekNext() == '*' {
				s.start()

				s.advance()
				s.advance()

				for !s.isAtEnd() && (s.peek() != '*' || s.peekNext() != '/') {
					if s.peek() == '\n' {
						s.line++
						s.column = 0
					}

					s.advance()
				}

				if !s.isAtEnd() {
					s.advance()

					if !s.isAtEnd() {
						s.advance()
					}
				}

				return s.make(Comment)
			} else {
				return PositionedToken{}
			}

		default:
			return PositionedToken{}
		}
	}
}

func (s *Scanner) peek() uint8 {
	if s.isAtEnd() {
		return '\000'
	}

	return s.text[s.currentI]
}

func (s *Scanner) peekNext() uint8 {
	if s.isAtEnd() {
		return '\000'
	}

	return s.text[s.currentI+1]
}

func (s *Scanner) advance() uint8 {
	s.currentI++
	s.column++

	return s.text[s.currentI-1]
}

func (s *Scanner) isAtEnd() bool {
	return s.currentI >= len(s.text)
}

func (s *Scanner) make(kind TokenKind) PositionedToken {
	return PositionedToken{
		Token: Token{
			Kind:   kind,
			Lexeme: s.text[s.startI:s.currentI],
		},
		Pos: core.Pos{
			Line:   s.startLine,
			Column: s.startColumn,
		},
	}
}

func (s *Scanner) start() {
	s.startI = s.currentI

	s.startLine = s.line
	s.startColumn = s.column
}

func (s *Scanner) error(msg string) PositionedToken {
	return PositionedToken{
		Token: Token{
			Kind:   Error,
			Lexeme: msg,
		},
		Pos: core.Pos{
			Line:   s.line,
			Column: s.column,
		},
	}
}

func isAlpha(c uint8) bool {
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_'
}

func isDigit(c uint8) bool {
	return c >= '0' && c <= '9'
}

func isHex(c uint8) bool {
	return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')
}

func isBinary(c uint8) bool {
	return c == '0' || c == '1'
}
